designs/node-ownership.md

```diff
@@ -0,0 +1,225 @@
+
+***Relevant GH Issues***
+
+1. [Karpenter ignores extra labels coming from kubelet-extra-args in launchtemplate](https://github.com/aws/karpenter/issues/1681)
+
+## Solutions
+
+Ultimately, the issues listed above involving node creation aren’t exclusive to just creation of the node. The Node is a fundamental component that every Kubernetes user takes some dependency on. As such, it’s reasonable to assume that users have written logic for scheduling/health checks on their nodes that have direct dependency on how a node looks and acts. Fundamentally, it’s a dangerous proposition to mutate the node object in ways that a standard Kubernetes user wouldn’t expect.
+
+As a result, all solutions that solve the node creation problem should:
+
+1. Capture in-flight capacity in a way so that we can quickly complete Karpenter’s provisioning loops and continue to start the next round of provisioning while having a consistent view of cluster state
+2. Completely remove (or, at a minimum, limit) the mutation of the Node so that we don’'t alter the Node in ways that the user doesn't expect
+3. Have little to no impact on cluster performance, including memory usage in etcd and query performance on the kube-apiserver.
+
+Fundamentally, if we do not capture in-flight capacity in the Node object, we need some other place to capture capacity so that we can fulfill #1 listed above. This necessitates some kind of “store” to keep in-flight node data, whether this is an in-memory store or the Kubernetes api-server (etcd) store. The two options presented below highlight both options, including their tradeoffs related to the fundamentals mentioned above.
+
+### Kubernetes CRD Object Store (Machine CRD)
+
+Karpenter will no longer create node objects or launch instances as part of the provisioning loop, but, instead, will create Machine CRs at the completion of the provisioning loop. This machine CR will then be picked up by a separate controller that will launch capacity based on the requirements passed from the provisioning loop and will resolve the static values from the **CreateFleet** response into its status.
+
+#### Impact
+
+1. Introduction of the Machine CR as a representation of a “scheduling decision” to launch by the CloudProvider
+2. Migration of Deprovisioning mechanisms from watching and monitoring the Node object to watching and monitoring the Machine CR
+3. Migration of the Node termination finalizer to Machine to orchestrate the cordoning, draining, and deletion
+4. Required Garbage Collection of any Machines that don’t have Nodes mapped to them
```

Comment:
Similar to what CCM does, I would suggest htat Karpenter GC Machine objects by watching the cloud API. E.g when an AWS EC2 instance has been terminated, the corresponding Machine CR is removed. Any Machine would exist without a corresponding Node before the Node's creation.

However, I suspect this line is also about terminating the underlying instance when a user terminates the Node object. In which case, you may want to do both.

Reply:
I think I get your point here. Machine owns the instance and is solely responsible for it. From a philosophical perspective, it makes sense that the Machine should only care about the instance it owns and nothing else.

Reply:
Updated the RFC based on your suggestion.

Reply:
Updated the RFC based on your suggestion.